src/borg/remote.py

```diff
@@ -726,6 +727,16 @@ def handle_error(unpacked):
                             return
                 except KeyError:
                     break
+            if cmd == 'barrier' and not self.ignore_responses:
+                # start processing barrier command(s) when we have received ALL outstanding async responses.
+                # raise one async exception per barrier command, in arbitrary order:
+                try:
+                    msgid, unpacked = self.async_responses.popitem()
+                except KeyError:
+                    # nothing left, don't raise exception, just return (None)
+                    return
+                else:
+                    handle_error(unpacked)
             if self.to_send or ((calls or self.preload_ids) and len(waiting_for) < MAX_INFLIGHT):
```

Comment:
I think the current code triggers this, so w_fds is set and pops one call in the write code path later, that makes this code even have more special cases. One way would be to ensure that calls is empty (by skipping the @api decorator and calling call_many explicitly below) from the start.